apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java

```diff
@@ -172,6 +319,83 @@ public void shutdown(Duration waitTime) throws ProcessingException {
    */
   @Override
   public void close() throws Exception {
-    writerPool.close();
+    log.info("close called");
+    shutdown(Duration.ofMinutes(2));
+    log.info("close complete");
+  }
+
+  private void throwIfErrorPresent() throws ProcessingException {
+    final var error = errors.getAndSet(null);
+    if (error != null) {
+      throw new ProcessingException(error, 0);
+    }
+  }
+
+  /**
+   * Performs necessary state updates based on the outcome of a batch write. Increments the
+   * uncollected processed messages count, records the error (if any) or (if successful) updates the
+   * set of written sequence numbers, then tells the publisher to allow more messages to be emitted.
+   *
+   * @param result the details of a completed batch from {@link ClaimWriter}
+   */
+  private void processBatchResult(BatchResult<TMessage> result) {
+    unreportedProcessedCount.addAndGet(result.getProcessedCount());
+    if (result.getError() != null) {
+      errors.compareAndSet(null, result.getError());
+    } else {
+      for (ApiMessage<TMessage> message : result.getMessages()) {
+        sequenceNumbers.removeWrittenSequenceNumber(message.getSequenceNumber());
+      }
+    }
+    publisher.allow(result.getMessages().size());
+  }
+
+  /**
+   * Remains true until shutdown is triggered.
+   *
+   * @return true if shutdown has not ben triggered
+   */
+  private boolean isRunning() {
+    return running.get();
+  }
+
+  /**
+   * Waits for countdown latch to reach zero so that rest of shutdown can proceed. The claim and
+   * sequence number writers decrement the latch as they complete so once it reach zero we know they
+   * are finished. We retry a few times in case of interrupts to resolve problems with spurious
+   * interrupts during cancellation. We can't stop the shutdown process completely so we return even
+   * if the latch doesn't reach zero before the timeout elapses.
+   *
+   * @param waitTime how long to wait for the latch
+   * @throws Exception if any exception caused the wait to fail
+   */
+  private void waitForLatch(Duration waitTime) throws Exception {
+    final long startMillis = System.currentTimeMillis();
+    InterruptedException error = null;
+    boolean successful = false;
+    for (int i = 1; i <= 10; ++i) {
+      try {
```

Comment:
wondering if we should we make this a constant explaining what this is ?
its hard to know why we would use 10 vs. any other number here

Reply:
I have refactored the method to remove the need for this counter.